src/game/common/thing/thingfactory.cpp

```diff
@@ -13,15 +13,167 @@
+ThingFactory::ThingFactory() : m_firstTemplate(nullptr), m_nextTemplateID(0) {}
+
+ThingFactory::~ThingFactory()
+{
+    Free_Database();
+}
+
+ThingTemplate *ThingFactory::New_Template(const Utf8String &name)
+{
+    ThingTemplate *new_template = NEW_POOL_OBJ(ThingTemplate);
+    const ThingTemplate *default_template = Find_Template("DefaultThingTemplate", false);
+
+    if (default_template) {
+        *new_template = *default_template;
+        new_template->Set_Copied_From_Default();
+    }
+
+    new_template->Friend_Set_Template_ID(m_nextTemplateID);
+    m_nextTemplateID++;
+
+    if (!m_nextTemplateID) {
+        captainslog_debug("m_nextTemplateID wrapped to zero");
+    }
+
+    new_template->Friend_Set_Template_Name(name);
+    Add_Template(new_template);
+    return new_template;
+}
+
+ThingTemplate *ThingFactory::New_Override(ThingTemplate *thing_template)
+{
+    captainslog_dbgassert(thing_template, "NULL 'parent' thing template");
+    captainslog_dbgassert(Find_Template(thing_template->Get_Name(), true),
+        "Thing template '%s' not in master list",
+        thing_template->Get_Name());
+
+    ThingTemplate *new_template = NEW_POOL_OBJ(ThingTemplate);
+    ThingTemplate *overrid = (ThingTemplate *)thing_template->Get_Final_Override();
+    *new_template = *overrid;
+    new_template->Set_Copied_From_Default();
+    new_template->Set_Is_Allocated();
+    overrid->Set_Next(new_template);
+    return new_template;
+}
+
+void ThingFactory::Free_Database()
+{
+    while (m_firstTemplate) {
+        ThingTemplate *t = m_firstTemplate;
+        m_firstTemplate = t->Friend_Get_Next_Template();
+        t->Delete_Instance();
+    }
+
+    m_templateMap.clear();
+}
+
+void ThingFactory::PostProcessLoad()
+{
+    for (ThingTemplate *i = m_firstTemplate; i; i = i->Friend_Get_Next_Template()) {
+        i->Resolve_Names();
+    }
+}
+
+void ThingFactory::Reset()
+{
+    ThingTemplate *t = m_firstTemplate;
+
+    while (t) {
+        bool first = false;
+        ThingTemplate *t2 = t->Friend_Get_Next_Template();
+
+        if (t == m_firstTemplate) {
+            first = true;
+        }
+
+        Utf8String str(t->Get_Name());
+        Overridable *o = t->Delete_Overrides();
+
+        if (!o) {
+            if (first) {
+                m_firstTemplate = t2;
+            }
+        }
+
+        if (!o) {
+            m_templateMap.erase(str);
+        }
+
+        t = t2;
+    }
+}
+
+void ThingFactory::Add_Template(ThingTemplate *tmplate)
+{
+    if (m_templateMap.find(tmplate->Get_Name()) != m_templateMap.end()) {
+        captainslog_debug("Duplicate Thing Template name found: %s", tmplate->Get_Name());
+    }
+
+    tmplate->Friend_Set_Next_Template(m_firstTemplate);
+    m_firstTemplate = tmplate;
+    m_templateMap[tmplate->Get_Name()] = tmplate;
+}
+
+const ThingTemplate *ThingFactory::Find_Template_By_ID(unsigned short id)
+{
+    for (ThingTemplate *i = m_firstTemplate; i; i = i->Friend_Get_Next_Template()) {

```

Comment:
again stick to one var name preferably `t` for all of the loops in this file.
Also != nullptr.

Reply:
Fixed.